File: WarmUp/searchclient/searchClientcpp/SearchClient.cpp
#include "SearchClient.h"
#include <string>
#include <list>
#include <unordered_map>
#include <regex>
#include <iostream>
#include <iterator>
#include "Strategy.h"
#include "Command.h"

using std::stringstream;
using std::istream;
using std::vector;
using std::string;
using std::getline;
using std::pair;
using std::list;

bool isAgent(char c) {
	return ('0' <= c && c <= '9');
}

bool isBox(char c) {
	return ('A' <= c && c <= 'Z');
}

bool isGoal(char c) {
	return ('a' <= c && c <= 'z');
}

SearchClient::SearchClient()
{
	std::unordered_map<char, std::string> colors;
	std::string line;

	std::regex color_regex("^[a-z]+:\\s*[0-9A-Z](\\s*,\\s*[0-9A-Z])*\\s*$");
	std::smatch match;


	while (getline (std::cin, line) && !line.compare("") && std::regex_match(line, match, color_regex)) {
		std::cerr << "Hi!";
		printf("%s\n", line.c_str());
		stringstream ss(line);
		string color;
		getline(ss, color, ':');

		string c;
		while (getline(ss, c, ',')) {
			c.erase(remove(c.begin(), c.end(), ' '), c.end());
			colors.insert(pair<char,string>(c.c_str()[0], color));
		}

	}

	int agents = 0;
	int boxes = 0;
	for (auto it = colors.begin(); it != colors.end(); ++it) {
		if (isAgent(it->first)) {
			agents++;
		}

		if (isBox(it->first)) {
			boxes++;
		}
	}

	list<string> rows;

	int cols = 0;
	do
	{
		//moreLines = !serverMessages.eof();
		if (line.length() > cols) {
			cols = line.length();
		}

//		std::cerr << line;
		rows.push_back(line);
		getline(std::cin, line);
	} while (!std::cin.eof() && (line.length() != 0));

	std::cerr << "Agents: " << agents <<
		"\nBoxes: " << boxes <<
		"\nDim: [" << cols << "," << rows.size() << "]\n";

//The max number of a row index is the amount of columns.
	Node::MAX_ROW = rows.size();
	Node::MAX_COL = cols;

	int size = rows.size();

	Node::walls.resize(rows.size()*cols, false);
	Node::goals.resize(rows.size()*cols, '\0');


	initialState = new Node();
	initialState->boxes = vector<char>(rows.size()*cols, '\0');
	//std::cerr << "cols is equal to " << cols << "\n";
	for (int y = 0; y < size; y++){
	//	std::cerr << "Y is equal to " << y << "\n";
		string row = rows.front();
		rows.pop_front();
		for (int x = 0; x < row.length(); x++){
	//		std::cerr << "X is equal to " << x << "\n";
			char chr = row[x];


			if (chr == '+'){
				Node::walls[y + x*size] = true;

			} else if (isAgent(chr)){
				initialState->agentRow = x;
				initialState->agentCol = y;
			} else if (isBox(chr)){
				initialState->boxes[y+x*size] = chr;
			} else if (isGoal(chr)){
				Node::goals[y+x*size] = chr;
			} else if (chr == ' '){
				//Do nothing
			} else {
	//			std::cerr << "Error, read invalid level character: [" << chr << "]" ;
			}
		}
	}
	std::cerr << initialState->toString();

}


SearchClient::~SearchClient()
{
}


int main(int argc, char * argv[]){
	//stringstream
	char buffer[200];
	int x;
	for (int i = 0; i < argc; i++){
		std::cerr << argv[i] << "\n";
	}

	SearchClient client = SearchClient();

	std::string strat = std::string(argv[0]);
	std::cerr << strat;
	std::cerr << "SearchClient started\n";

	StrategyBFS strategy = StrategyBFS();
	std::cerr << "Defaulting to BFS search. Use arguments -bfs, -dfs, -astar, -wastar, or -greedy to set the search strategy.\n";

	std::list<Node *> solution;
	solution = client.search(&strategy);
	std::cerr << "\nSummary for " << strategy.toString() << ".\n";
	std::cerr << "Found solution of length " << solution.size() << ".\n";
	//std::cerr << strategy.searchStatus();

	for (Node * n : solution) {
		//std::cerr << "Printing solution1";
		if (n == NULL){
			std::cerr << "Something is null";
		}
		std::string act = n->action->to_string();
		//std::cerr << n->action->to_string();
		//std::cerr << "Printing solution2";
		std::cerr << act;
		std::cout << act;
		std::cout << "\n";
		std::string response;
		//std::cerr << "Printing solution3";
		std::getline(std::cin, response);
		//std::cerr << "Printing solution4";
		if (response.find(std::string("false")) != std::string::npos) {
			sprintf(buffer, "Server responsed with %s to the inapplicable action: %s\n", response, act);
			std::cerr << std::string(buffer);
			sprintf(buffer, "%s was attempted in \n%s\n", act, n->toString());
			std::cerr << std::string(buffer);
			break;
		}
		}
		return 0;
	}

	std::list<Node *> SearchClient::search(Strategy * strategy) {
		char buffer[100];
		sprintf(buffer, "Search starting with strategy %s.\n", strategy->toString().c_str());
		std::string s = std::string(buffer);
		std::cerr << s;
		strategy->addToFrontier(this->initialState);
		int iterations = 0;

		while (true) {
			//std::cerr << "Iteration!\n";
			if (iterations == 1000) {
				std::cerr << strategy->searchStatus();
				iterations = 0;
			}

			if (strategy->frontierIsEmpty()) {
				//Empty list
				return std::list<Node *>();
			}

			Node * leafNode = strategy->getAndRemoveLeaf();

			if (leafNode->isGoalState()) {
				std::cerr << "Goal!\n";
				std::cerr << leafNode->toString();
				return leafNode->extractPlan();
			}

			strategy->addToExplored(leafNode);
			std::vector<Node *> nodes = leafNode->getExpandedNodes();
			//std::cerr << "Number of new elements: " << nodes.size() << "\n";
			for (Node * n : nodes) {
				if (!strategy->isExplored(n) && !strategy->inFrontier(n)) {
					strategy->addToFrontier(n);
				}
			}
			iterations++;
		}
	}

File: oldBackup/WarmUp/searchclient/searchclient/SearchClient.java
package searchclient;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;

import searchclient.Memory;
import searchclient.Strategy.*;
import searchclient.Heuristic.*;

public class SearchClient {
	public Node initialState;


    public SearchClient(BufferedReader serverMessages) throws Exception {
		// Read lines specifying colors
		String line = serverMessages.readLine();
		if (line.matches("^[a-z]+:\\s*[0-9A-Z](\\s*,\\s*[0-9A-Z])*\\s*$"))
		{
			System.err.println("Error, client does not support colors.");
			System.exit(1);
		}

		int row = 0;
		boolean agentFound = false;
		this.initialState = new Node(null);
		// For loading use arraylist as size is not yet known.
		ArrayList<Box> boxes = new ArrayList<>();
		ArrayList<Agent> agents = new ArrayList<>();
		ArrayList<Goal> goals = new ArrayList<>();

		while (!line.equals(""))
		{
			for (int col = 0; col < line.length(); col++)
			{
				char chr = line.charAt(col);

				if(line.length() > Node.MAX_COL)
				    Node.MAX_COL = line.length();

				if (chr == '+')
				{ // Wall.
					Node.walls[row][col] = true;
				}
				else if ('0' <= chr && chr <= '9')
				{ // Agent.
					agents.add(new Agent((int) chr, new Pair<>(row, col)));
				}
				else if ('A' <= chr && chr <= 'Z')
				{
					// Box.
					boxes.add(new Box(chr, new Pair<>(row, col)));
				}
				else if ('a' <= chr && chr <= 'z')
				{
					// Goal.
					goals.add(new Goal(chr, new Pair<>(row, col)));
				}
				else if (chr == ' ')
				{
					// Free space.
				}
				else
				{
					System.err.println("Error, read invalid level character: " + (int) chr);
					System.exit(1);
				}
			}
			line = serverMessages.readLine();
			row++;
		}
		Node.MAX_ROW = row;
		// Arraylists to arrays
		Node.goals = new Goal[goals.size()];
		Node.goals = goals.toArray(Node.goals);
		this.initialState.boxes = new Box[boxes.size()];
		this.initialState.boxes = boxes.toArray(this.initialState.boxes);
		this.initialState.agents = new Agent[agents.size()];
		this.initialState.agents = agents.toArray(this.initialState.agents);
	}

	public LinkedList<Node> Search(Strategy strategy) throws IOException {
		System.err.format("Search starting with strategy %s.\n", strategy.toString());
		strategy.addToFrontier(this.initialState);

		int iterations = 0;
		while (true) {
            if (iterations == 1000) {
				System.err.println(strategy.searchStatus());
				iterations = 0;
			}

			if (strategy.frontierIsEmpty()) {
				return null;
			}

			Node leafNode = strategy.getAndRemoveLeaf();

			

			if (leafNode.isGoalState()) {
				return leafNode.extractPlan();
			}

			strategy.addToExplored(leafNode);
			//System.err.println("Hashcode of agent: " + String.valueOf(leafNode.hashCode()));
			//System.err.println("Hej: " + leafNode.isGoalState());
			for (Node n : leafNode.getExpandedNodes()) { // The list of expanded nodes is shuffled randomly; see Node.java.
				if (!strategy.isExplored(n) && !strategy.inFrontier(n)) {
					strategy.addToFrontier(n);
				}
			}
			iterations++;
		}
	}

	public static void main(String[] args) throws Exception {
		BufferedReader serverMessages = new BufferedReader(new InputStreamReader(System.in));

		// Use stderr to print to console
		System.err.println("SearchClient initializing. I am sending this using the error output stream.");

		// Read level and create the initial state of the problem
		SearchClient client = new SearchClient(serverMessages);

        Strategy strategy;
        if (args.length > 0) {
            switch (args[0].toLowerCase()) {
                case "-bfs":
                    strategy = new StrategyBFS();
                    break;
                case "-dfs":
                    strategy = new StrategyDFS();
                    break;
                case "-astar":
                    strategy = new StrategyBestFirst(new AStar(client.initialState));
                    break;
                case "-wastar":
                    // You're welcome to test WA* out with different values, but for the report you must at least indicate benchmarks for W = 5.
                    strategy = new StrategyBestFirst(new WeightedAStar(client.initialState, 5));
                    break;
                case "-greedy":
                    strategy = new StrategyBestFirst(new Greedy(client.initialState));
                    break;
                default:
                    strategy = new StrategyBFS();
                    System.err.println("Defaulting to BFS search. Use arguments -bfs, -dfs, -astar, -wastar, or -greedy to set the search strategy.");
            }
        } else {
            strategy = new StrategyBFS();
            System.err.println("Defaulting to BFS search. Use arguments -bfs, -dfs, -astar, -wastar, or -greedy to set the search strategy.");
        }

		LinkedList<Node> solution;
		try {
			solution = client.Search(strategy);
		} catch (OutOfMemoryError ex) {
			System.err.println("Maximum memory usage exceeded.");
			solution = null;
		}

		if (solution == null) {
			System.err.println(strategy.searchStatus());
			System.err.println("Unable to solve level.");
			System.exit(0);
		} else {
			System.err.println("\nSummary for " + strategy.toString());
			System.err.println("Found solution of length " + solution.size());
			System.err.println(strategy.searchStatus());

			for (Node n : solution) {
				String act = n.action.toString();
				System.out.println(act);
				String response = serverMessages.readLine();
				if (response.contains("false")) {
					System.err.format("Server responsed with %s to the inapplicable action: %s\n", response, act);
					System.err.format("%s was attempted in \n%s\n", act, n.toString());
					break;
				}
			}
		}
	}
}
